pkg/resource/api.go

```diff
@@ -111,6 +111,10 @@ func (a *APIPatchingApplicator) Apply(ctx context.Context, o runtime.Object, ao
 		return errors.New("cannot access object metadata")
 	}
 
+	if m.GetName() == "" && m.GetGenerateName() != "" {
+		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
+	}
```

Comment:
Nitpick: Could we add this to the `kerrors.IsNotFound(err)` check below?

```
if kerrors.IsNotFound(err) || (err != nil && m.GetGenerateName() != "") {
		// TODO(negz): Apply ApplyOptions here too?
		return errors.Wrap(a.client.Create(ctx, o), "cannot create object")
}
```

Reply:
But the error is `name is required` rather than `resource not found`

Reply:
Right, this would check if the error is `IsNotFound`, and if not, would check if `m.GetGenerateName()` was populated. If either were true it would attempt to create the object (the same thing you do above). If this is a specific error type then you could also replace the `err != nil` with something like `kerrrors.IsNameRequred(err)`

Reply:
Yeah. But it feels awkward to make the `Get` call when you can be 100% sure that it's going to fail by checking `m.GetName() == ""`.

Reply:
Fair. I am fine either way, just seemed like some duplication of code :+1:

Reply:
Is it worth making this check a function in `pkg/meta`? `meta.PendingName` or something?

Reply:
@negz Do you have a another use-case in mind?

Reply:
Not really; happy to just leave this as it is. We can always move it out if we find other use cases.

Reply:
Yea, it's just I'm trying to be a bit wary of generic things with one consumer.